src/Microsoft.AspNetCore.Blazor/Components/ChangeDetection.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Blazor.Components
+{
+    internal class ChangeDetection
+    {
+        public static bool MayHaveChanged<T1, T2>(T1 oldValue, T2 newValue)
+        {
+            var oldIsNotNull = oldValue != null;
+            var newIsNotNull = newValue != null;
+            if (oldIsNotNull != newIsNotNull)
+            {
+                return true; // One's null and the other isn't, so different
+            }
+            else if (oldIsNotNull) // i.e., both are not null (considering previous check)
+            {
+                var oldValueType = oldValue.GetType();
+                var newValueType = newValue.GetType();
+                if (oldValueType != newValueType            // Definitely different
+                    || !IsKnownImmutableType(oldValueType)  // Maybe different
+                    || !oldValue.Equals(newValue))          // Somebody says they are different
```

Comment:
It's because conceptually, the values we're comparing aren't really `oldValue` and `newValue`. The values we want to compare are *the state reachable transitively from `oldValue` at some unknown point in the past* versus *the state reachable transitively from `newValue` now*. The purpose of the check is to determine whether there's any chance that the deeply-reachable state is different now versus before.

If this was public API, we'd have to be much clearer about the semantics of this check and do some naming review to try to communicate this better. We could still change the name, but since it's `internal`, it's not as urgent.

With these semantics in mind, the only case where we can return `true` is for types known to be immutable. For, say, `object`, it's not enough for `oldValue.Equals(newValue)`, because in most cases `oldValue` and `newValue` will be handles to the same object instance but that doesn't tell us anything about whether it's mutated.

It's unfortunate that .NET doesn't have any native concept of (im)mutability. That limits us to hard-coding a set of commonly-used known-immutable types.